// kick_live_ws/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EChatMessageEventType = exports.LIVESTREAM = exports.CURRENT_VIEWER = exports.CHANNEL_V1 = exports.CHANNEL = exports.scrapeWebsite = exports.WebSocketConnection = void 0;
var websocket_1 = require("./websocket");
Object.defineProperty(exports, "WebSocketConnection", { enumerable: true, get: function () { return websocket_1.WebSocketConnection; } });
var userAPI_1 = require("./fetch/userAPI");
Object.defineProperty(exports, "scrapeWebsite", { enumerable: true, get: function () { return userAPI_1.scrapeWebsite; } });
var kick_1 = require("./api-routes/kick");
Object.defineProperty(exports, "CHANNEL", { enumerable: true, get: function () { return kick_1.CHANNEL; } });
Object.defineProperty(exports, "CHANNEL_V1", { enumerable: true, get: function () { return kick_1.CHANNEL_V1; } });
Object.defineProperty(exports, "CURRENT_VIEWER", { enumerable: true, get: function () { return kick_1.CURRENT_VIEWER; } });
Object.defineProperty(exports, "LIVESTREAM", { enumerable: true, get: function () { return kick_1.LIVESTREAM; } });
var websocketTypes_1 = require("./types/websocketTypes");
Object.defineProperty(exports, "EChatMessageEventType", { enumerable: true, get: function () { return websocketTypes_1.EChatMessageEventType; } });
